utils/ensemble_tools.py

```diff
@@ -0,0 +1,166 @@
+    scan_points = {}
+    list_all = []
+    for parameter in parameter_dict.keys():
+        param_value = parameter_dict[parameter][0]
+        scan_points[parameter] = []
+        for i in range(n_dirs):
+            scan_points[parameter].append(param_value)
+            list_all.append(f"{parameter}_{i}")
+            param_value += (
+                parameter_dict[parameter][1] - parameter_dict[parameter][0]
+            ) / float(n_dirs - 1)
+
+    # Create the combinations of these parameters.
+    combinations = list(itertools.combinations(list_all, len(parameter_dict.keys())))
+    all_configs = []
+    for comb in combinations:
+        record_comb = True
+        for i in range(len(comb) - 1):
+            search_term = comb[i][: len(comb[i]) - comb[i][::-1].find("_")]
+            for j in range(i + 1, len(comb)):
+                if comb[j].startswith(search_term):
+                    record_comb = False
+                    break
+        if record_comb:
+            all_configs.append(comb)
+
+    # If destructive, let's delete the whole tree if it already exists
+    if destructive and os.path.isdir(sweep_path):
+        shutil.rmtree(sweep_path)
+    # Check whether the copy directory exists
+    copy_files = os.path.isdir(copy_dir)
+    # Now loop over each combination, create a directory, copy the
+    # configs and conditions, and edit the configuation to contain
+    # the correct parameters.
+    for comb in all_configs:
+        dir_name = "-".join(comb)
+        dir_path = Path(sweep_path) / "SWEEP" / dir_name
+        make_directory(dir_path, destructive)
+        if copy_files:
+            copy_dir_contents(dir_path, copy_dir)
+        # Create parameter map for conditions encoding
+        parameters = {}
+        for parameter in parameter_dict.keys():
+            param_index = [item for item in comb if item.startswith(parameter)][
+                0
+            ].split("_")[-1]
+            parameters[parameter] = scan_points[parameter][int(param_index)]
+        if edit_file:
+            encode_conditions_file(Path(dir_path) / edit_file, parameters)
+
+
+def copy_dir_contents(dir_path, copy_dir):
+    """Copy the contents of copy_dir -> dir_path"""
+
+    shutil.copytree(copy_dir, dir_path, dirs_exist_ok=True)
+
+
+def make_directory(directory_name, destructive):
+    """Make the directory tree at directory_name."""
+
+    if os.path.isdir(directory_name):
+        if destructive:
+            shutil.rmtree(directory_name)
+        else:
+            msg = "Path already exists and not in destructive mode"
+            raise FileExistsError(msg)
+    os.makedirs(directory_name)
+
+
+def encode_conditions_file(in_file_name, param_dict):
+    """Encode a configuration file with a dict of input values"""
+
+    for param_name in param_dict.keys():
+        edit_parameter(in_file_name, param_name, param_dict[param_name])
+
+
+def edit_parameter(in_file_name, param, val):
```

Comment:
If rather than passing a singe `param`, `val` pair to `edit_parameter` you instead passed a parameter to value dict then you could parse the XML file only once - something like

```Python
def edit_parameters(in_file_name, parameter_overrides):
    """Edit one or more parameters in conditions file."""
    data = ET.parse(in_file_name)
    root = data.getroot()
    parameters = root.find("CONDITIONS").find("PARAMETERS")
    for element in parameters.iter("P"):
        match = re.match(
            r"\s*(?P<key>\w+)\s*=\s*(?P<value>[0-9.-]+)\s*",
            element.text,
        )
        if match is None:
            raise ValueError(f"Invalid parameter specification in base file: {element.text}")
        key = match.group("key")
        if key in parameter_overrides:
            element.text = f" {key} = {parameter_overrides[key]} "
    ET.indent(root, space="    ")
    data.write(in_file_name)
```
which would then mean we could get rid of `encode_conditions_file` and use `edit_parameters` in place of both the current `encode_condition_file` and `edit_parameter` calls.